Services/UICore/classes/class.ilCtrl.php

```diff
@@ -1,1890 +1,1116 @@
-        return $this->class_cid[$a_class];
-    }
 
-    private function cidClassUnknown($a_cid)
-    {
-        return (!isset($this->cid_class[$a_cid]) || $this->cid_class[$a_cid] == "");
-    }
+        // append a csrf token if the command is considered
+        // unsafe or the link is for form actions.
+        if ($is_post || !$this->isCmdSecure($cmd_class, $a_cmd)) {
+            global $DIC;
 
+            $token = new ilCtrlToken(
+                $DIC->database(),
+                $DIC->user()
+            );
 
-    /**
-     * Get class for class id after fetching and storing corresponding information, if necessary.
-     */
-    private function getClassForCid($a_cid)
-    {
-        if ($this->cidClassUnknown($a_cid)) {
-            $this->readCidInfo($a_cid);
+            $target_url = $this->appendParameterString(
+                $target_url,
+                self::PARAM_CSRF_TOKEN,
+                $token->getToken(),
+                $is_escaped
+            );
         }
-        if ($this->cidClassUnknown($a_cid)) {
-            throw new ilCtrlException("Cannot find class for cid " . $a_cid . ".");
-        }
-        return $this->cid_class[$a_cid];
-    }
 
-    private function fetchCallsOfClassFromCache($a_class, ilCachedCtrl $a_cached_ctrl)
-    {
-        foreach ($a_cached_ctrl->lookupCall($a_class) as $call) {
-            if ($call["child"] != "" && $this->callOfClassNotKnown($a_class, $call['child'])) {
-                $this->calls[$a_class][] = $call["child"];
-            }
+        if ($is_async) {
+            $target_url = $this->appendParameterString(
+                $target_url,
+                self::PARAM_CMD_MODE,
+                self::CMD_MODE_ASYNC,
+                $is_escaped
+            );
         }
-    }
 
-    /**
-     * Save class respective to $a_cid and store corresponding
-     * class calls for future reference.
-     *
-     * @param object $a_cid		cid
-     */
-    private function readCidInfo($a_cid)
-    {
-        if (isset($this->info_read_cid[$a_cid])) {
-            return;
+        if (!empty($a_anchor)) {
+            $target_url .= "#$a_anchor";
         }
 
-        $cached_ctrl = ilCachedCtrl::getInstance();
-        $cid_info = $cached_ctrl->lookupCid($a_cid);
-
-        if ($cid_info) {
-            $this->updateClassCidMap($cid_info['class'], $a_cid);
-            $this->fetchCallsOfClassFromCache($cid_info['class'], $cached_ctrl);
-            $this->info_read_class[$cid_info["class"]] = true;
-        }
-        
-        $this->info_read_cid[$a_cid] = true;
+        return $target_url;
     }
 
     /**
-     * Save classes respective to the class id's of a node and store corresponding
-     * class calls for future reference.
+     * This helper function wraps the deprecated UI functionality that
+     * modifies a URL target and "hacks into" existing HTML.
      *
-     * @param	string	$a_node
+     * (Tbh I don't really get that mechanism, but it stays for now.)
```

Comment:
Please dont remove this hook. Feel free to change it but i need the feature. Its quiet important to a lot of hacky things i do in some projects. I am using this as a after init hook.

Reply:
@PurHur I don't think that this method will stay forever, as the method `ilUIHookPluginGUI::getHTML()` is already marked as deprecated and should not be used in the future. The mechanism depends on this method and I'm not sure if there will be methods to replace this behaviour.

Reply:
Feel free to replace this method with another hook system. I am totally on your side that this should not be named "getHTML". It took me ages to find this hook in the first place. Change it as you like but remain the feature. All plugins now can listen on all events. That could be a direction

Reply:
Also how is deprecation handled in Ilias? I did not get a deprecation warning yet. 

@mjansenDatabay 
We use the typo3 system to deprecate methods. I did not see that in ilias. I think you are the right developer to push this topic in the right direction. Please see this link: 
https://docs.typo3.org/m/typo3/reference-coreapi/master/en-us/ApiOverview/Deprecation/Index.html#deprecate-functions-in-extensions

Reply:
@PurHur

> Also how is deprecation handled in Ilias? I did not got a deprecation warning yet. 


We mark the class/interface/method/you name it as `@deprecated`.

Reply:
Is there a possibility to trigger a real deprecation error.
Since i am managing a lot of code i cant read all of it all the time. This is a impossible task for me.  A deprecation notice would we send via https://sentry.io/ to me.

Reply:
I would really like that. And i would join the TA for that discussion if you like the idea behind that. This is not related to the headline of this PR.